diofant/polys/rootoftools.py

```diff
@@ -299,9 +303,52 @@ def _complexes_sorted(cls, complexes):
             return []
         cache = {}
 
+        def is_re_disjoint(u, v):
+            if u.conj != v.conj:
+                return True
+            else:
+                return u.bx <= v.ax or v.bx <= u.ax
+
+        res_cache = {}
+
+        def match_same_real_part(u, v):
+            assert u.domain == v.domain
+            dom = u.domain
+            if tuple(u.f) not in res_cache:
+                res_cache[tuple(u.f)] = dmp_resultant(dmp_permute(u.f1, [1, 0], 1, dom),
+                                                      dmp_permute(u.f2, [1, 0], 1, dom),
+                                                      1, dom)
+            if tuple(v.f) not in res_cache:
+                res_cache[tuple(v.f)] = dmp_resultant(dmp_permute(v.f1, [1, 0], 1, dom),
+                                                      dmp_permute(v.f2, [1, 0], 1, dom),
+                                                      1, dom)
+            g = dmp_gcd(res_cache[tuple(u.f)], res_cache[tuple(v.f)], 0, dom)
+            a, b = max(u.ax, v.ax), min(u.bx, v.bx)
+            roots = dup_isolate_real_roots(g, dom, inf=a, sup=b)
+            if len(roots) != 1:
+                return False
+            else:
+                a, b = roots[0][0]
+            if a == b:
```

Comment:
Beware, that part seems to be redundant.  But ``dup_count_complex_roots`` sometimes fails in case, where for x-component ``inf==sup``.

For example:
```
In [4]: dup_count_complex_roots([1, 0, 1], ZZ_python, inf=0, sup=0)
Out[4]: 2

In [5]: dup_count_complex_roots([1, 0, 1], ZZ_python, inf=(0, -1), sup=0)
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-5-d0b6fcdb6ef6> in <module>()
----> 1 dup_count_complex_roots([1, 0, 1], ZZ_python, inf=(0, -1), sup=0)

~/src/diofant/diofant/polys/rootisolation.py in dup_count_complex_roots(f, K, inf, sup, exclude)
   1303     Q_L4 = _intervals_to_quadrants(I_L4, f1L4F, f2L4F, t, v, F)
   1304 
-> 1305     T = _traverse_quadrants(Q_L1, Q_L2, Q_L3, Q_L4, exclude=exclude)
   1306 
   1307     return _winding_number(T, F)

~/src/diofant/diofant/polys/rootisolation.py in _traverse_quadrants(Q_L1, Q_L2, Q_L3, Q_L4, exclude)
   1181                 rules.append((_rules_ambiguous[qq], corners[(j, i)]))
   1182             else:  # pragma: no cover
-> 1183                 raise NotImplementedError("3 element rule (corner): " + str(qq))
   1184 
   1185         q1, k = Q[0], 1

NotImplementedError: 3 element rule (corner): ('OO', 'OO', 'A1')
```
```

Reply:
The corners are particularly tricky. They may lead to degenerate intervals. Maybe they should be extracted and treated separately.

Reply:
I've added a rule, that removes need in this workaround.  Not sure if it's correct, yet.

BTW, it seems that only line segments, which are parallel to imaginary axis - poses problems.  Other cases of degenerate isolation rectangles works ok.